File: JavaScript/Concurrency/BoundedBufferDemo.js
'use strict';

class BoundedBuffer {
    constructor(capacity) {
        this.queue = [];
        this.capacity = capacity;
        this.producerWaiters = [];
        this.consumerWaiters = [];
    }

    async put(item) {
        while (this.queue.length === this.capacity) {
            await new Promise((resolve) => this.producerWaiters.push(resolve));
        }
        this.queue.push(item);
        this._notifyConsumer();
    }

    async take() {
        while (this.queue.length === 0) {
            await new Promise((resolve) => this.consumerWaiters.push(resolve));
        }
        const item = this.queue.shift();
        this._notifyProducer();
        return item;
    }

    _notifyConsumer() {
        const waiter = this.consumerWaiters.shift();
        if (waiter) waiter();
    }

    _notifyProducer() {
        const waiter = this.producerWaiters.shift();
        if (waiter) waiter();
    }
}

async function produce(buffer) {
    for (let i = 0; i < 10; i++) {
        await buffer.put(i);
        console.log(`Produced: ${i}`);
    }
}

async function consume(buffer) {
    for (let i = 0; i < 10; i++) {
        const item = await buffer.take();
        console.log(`Consumed: ${item}`);
    }
}

async function main() {
    const buffer = new BoundedBuffer(5);
    await Promise.all([produce(buffer), consume(buffer)]);
}

main();

File: Python/Concurrency/BoundedBufferDemo.py
import threading
from collections import deque
from typing import Deque, Generic, TypeVar

T = TypeVar("T")


class BoundedBuffer(Generic[T]):
    def __init__(self, capacity: int) -> None:
        self._queue: Deque[T] = deque()
        self._capacity = capacity
        self._lock = threading.Lock()
        self._producer_cond = threading.Condition(self._lock)
        self._consumer_cond = threading.Condition(self._lock)

    def put(self, item: T) -> None:
        with self._producer_cond:
            while len(self._queue) == self._capacity:
                self._producer_cond.wait()
            self._queue.append(item)
            self._consumer_cond.notify()

    def take(self) -> T:
        with self._consumer_cond:
            while not self._queue:
                self._consumer_cond.wait()
            item = self._queue.popleft()
            self._producer_cond.notify()
            return item


def produce(buffer: "BoundedBuffer[int]") -> None:
    for i in range(10):
        buffer.put(i)
        print(f"Produced: {i}")


def consume(buffer: "BoundedBuffer[int]") -> None:
    for _ in range(10):
        item = buffer.take()
        print(f"Consumed: {item}")


def main() -> None:
    buffer: BoundedBuffer[int] = BoundedBuffer(5)

    t1 = threading.Thread(target=produce, args=(buffer,))
    t2 = threading.Thread(target=consume, args=(buffer,))
    t1.start()
    t2.start()
    t1.join()
    t2.join()


if __name__ == "__main__":
    main()
